### ISSUE TITLE
Indexing is not happening in Real Time for Simple Property Deletion 

### ISSUE THREAD

**Issue by singhik** [NCIEVS/nci-protege5#434]
Index is not happening in real time for Deletion of Simple Properties(f.e. Editor_Note) .

Attached is the screen shot document for reference ....
[Indexing_Issue.docx](https://github.com/NCIEVS/nci-protege5/files/3740766/Indexing_Issue.docx)

**Comment by singhik** [comment]
@bdionne : Hi, I am still not able to see Indexing happen Automatically post deletion of a property , Unless Index is Run Manually, the Deleted Property 'Search' seems to work .Document for reference is attached ....

[Issue 434.docx](https://github.com/NCIEVS/nci-protege5/files/4505524/Issue.434.docx)

**Comment by bdionne** [comment]
this appears to be a lower level issue, not readily reproducible, needs more research

**Comment by NCIEVS** [comment]
May have already been fixed on the autoupdate fix in 3.5.3

**Comment by singhik** [comment]
@bdionne  : This is still not fixed, auto updater is not indexing deleted properties unless indexing is manually Run . 

Able to search with a Pty (Editor_Note), which has been deleted : 

![image](https://user-images.githubusercontent.com/36314371/129593319-dac3b2b9-3bfc-4531-a7b4-fd2b233f1d43.png)

![image](https://user-images.githubusercontent.com/36314371/129593346-fc680078-9d95-481d-adcd-c7a126e958ea.png)

**Comment by bdionne** [comment]
@singhik I"m curious, after you delete the property do you see a messages on the console or in the log that looks like this?

````
2021-08-16 12:18:38.786 [pool-1-thread-1] INFO  SearchTabManager    Updating index with 1 change(s)
````

**Comment by singhik** [comment]
@bdionne : Hi, yes, i do get that message in the logs post delete : 

2021-08-16 12:32:59.856 [AWT-EventQueue-0] WARN  Injector    No files found for META-INF/services/org.semanticweb.owlapi.model.OWLOntologyIRIMapper
2021-08-16 12:32:59.857 [AWT-EventQueue-0] INFO  Injector    Injecting values [[]] on method public void uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.setIRIMappers(java.util.Set).
2021-08-16 12:33:02.929 [pool-1-thread-1] INFO  SearchTabManager    Updating index with 1 change(s)

**Comment by bdionne** [comment]
I suspect what's going on might be `fsync` on the windows machines. I never see this now on `OS X`, it works fine. When the changes occur there is a call to `commit()` in Lucene, which syncs any buffered changes to disk. In some OSes these writes are buffered for performance. When the `commit()` occurs we also mark the index dirty so that the next query will open the index for reading again.

**Comment by singhik** [comment]
@bdionne : Ok, i am fine with closing it as it's a low priority ticket and indexing works fine for creation ,only for deletion it needs to be run manually. 
@NCIEVS : Please advice .

**Comment by bdionne** [comment]
@singhik @NCIEVS sees the same problem on Linux, which makes me think it's not an `fsync` issue, I'll continue to look at the code and see if I can figure it out. Let's keep the ticket open and if I can't find it we can just move it to the next release.

**Comment by bdionne** [comment]
@singhik I pushed a 1.2.5 version of the LQT to the top level, if you want to try it point to my autoupdater. It definitely fixes the issue I outlined in email, and has a possible fix for the other issue. There may only be really one bug, the on eI identified with respect to the order of changes. It's best to test this by first bringing up the app, doing a manual index, close the app and then bring it up again. That should clean out any bad indices.

**Comment by singhik** [comment]
@NCIEVS :Hi, the flow which works for auto-indexing for deletion of properties is : 

> Delete a property, Build Index **manually** , perform Search with the deleted property , restart and again perform search with deleted property .

The deleted property is not searchable .

**Comment by singhik** [comment]
@bdionne @NCIEVS  : Moving this to 3.9.0 Milestone as this ticket works for NCIEVS in Windows but does not work for me, with same configurations, plugins  & server ......

**Comment by bdionne** [comment]
I'll keep looking into this one, it's starting to bug me (no pun intended)

**Comment by bdionne** [comment]
This likely relates to #547 and can possibly be fixed using an additional thread

**Comment by singhik** [comment]
@bdionne : This still does not work in real-time for Property Deletion on Windows Platform, not until manual indexing is run .

**Comment by bdionne** [comment]
@singhik - I more or less know what you tested, but please enumerate the exact steps you followed, if possible. Include the code of the class you tested, the property deleted, and details of the search query used to see if it was deleted.

Thanks

@NCIEVS Have you seen this fail on Windows?

**Comment by singhik** [comment]
@bdionne  : Pfb : 

Adding a Property to a Concept 'LE Cell' : 

![image](https://user-images.githubusercontent.com/36314371/163078260-2992b44c-d109-47cf-be1e-d54f536671a0.png)

Deleting the same and performing Search : 

![image](https://user-images.githubusercontent.com/36314371/163078382-3a3b08f3-b768-48ed-89f1-efe95ecf6ce4.png)
Search is able to find the Concept after Deleting the Property : 

![image](https://user-images.githubusercontent.com/36314371/163078406-51a0b1c6-96d3-4547-a69f-99a9f1a17f58.png)

**Comment by singhik** [comment]
@bdionne  : No, it fails, after deletion of property, that Search still works, same failure as it has been since this ticket started.

**Comment by NCIEVS** [comment]
@bdionne @singhik , it's not working now for me in linux, for either simple or complex properties.  The search still returns the concept after the target of the search has been deleted.

**Comment by bdionne** [comment]
I'm seeing the same thing. I think this is related to #552

**Comment by singhik** [comment]
@NCIEVS : THanks,  @bdionne  : Even Simple Property Addition is also not working for me on 1.3.9  . Steps : 

1. Add a Simple Property f.e. 'Editor_Note' to a concept 

![image](https://user-images.githubusercontent.com/36314371/163605103-e242f114-4cfc-460e-985c-e317f2183ccd.png)

2. Perform Search with Editor_Note :  Unable to Search the Concept 

![image](https://user-images.githubusercontent.com/36314371/163605188-73cd0731-7415-4388-a5bd-939348044fe7.png)

**Comment by singhik** [comment]
@NCIEVS @bdionne : We can move this ticket to next Milestone Release, i am already in fourth iteration of testing, not feasible to perform another round with another Edit Tab plug-in

**Comment by bdionne** [comment]
This is related to #552 as noted above, and has a fix that I'm testing now. It's going to require new plugins for both EditTab and Lucene (already pushed), as well as a new 4.0.0-dev2 tag.

**Comment by bdionne** [comment]
This is fixed in [nci-edit-tab-1.4.0](https://github.com/NCIEVS/nci-edit-tab/releases/tag/nci-edit-tab-1.4.0). It requires [lucene-search-tab-1.3.0](https://github.com/NCIEVS/lucene-search-tab/releases/tag/lucene-search-tab-1.3.0) and the protege [4.0.0-dev2](https://github.com/NCIEVS/protege/releases/tag/4.0.0-dev2) build.

This also fixes #552 and #547

**Comment by singhik** [comment]
@bdionne @NCIEVS : Fails for Real Time Search post Deletion of Property, 

Adding a Property to a Concept : 

![image](https://user-images.githubusercontent.com/36314371/164993957-d2ddc235-9259-4b9e-b8f2-acf8eefefef8.png)

![image](https://user-images.githubusercontent.com/36314371/164993965-d2d24fe7-07d8-4f21-b16d-e4a129f38a5d.png)

Performing Search : 

![image](https://user-images.githubusercontent.com/36314371/164993978-aee5aa2b-42d8-4e29-921b-43b8f12de731.png)

Deleting above Property : 

![image](https://user-images.githubusercontent.com/36314371/164993992-a55afc36-32c5-4d5e-9b61-56332bd3953f.png)

Performing Search :  Search results still display the Concept after removing the Property 

![image](https://user-images.githubusercontent.com/36314371/164994032-b54103c9-a983-4f6e-9172-78b1279fb941.png)

**Comment by singhik** [comment]
This ticket is working on @NCIEVS 's machine on both Linux and Windows Platform, hence is closed
